progress/ansimeter.go

```diff
@@ -0,0 +1,197 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package progress
+
+import (
+	"fmt"
+	"time"
+	"unicode"
+
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+// ANSIMeter is a progress.Meter that uses ANSI escape codes to make
+// better use of the available horizontal space.
+type ANSIMeter struct {
+	label   []rune
+	total   float64
+	written float64
+	spin    int
+	t0      time.Time
+}
+
+// these are the bits of the ANSI escapes (beyond \r) that we use
+// (names of the terminfo capabilities, see terminfo(5))
+var (
+	// clear to end of line
+	clrEOL = "\033[K"
+	// make cursor invisible
+	cursorInvisible = "\033[?25l"
+	// make cursor visible
+	cursorVisible = "\033[?12;25h"
+	// turn on reverse video
+	enterReverseMode = "\033[7m"
+	// go back to normal video
+	exitAttributeMode = "\033[0m"
+)
+
+func getCol() int {
+	col, _, _ := terminal.GetSize(0)
+	if col <= 0 {
+		// give up
+		col = 80
+	}
+	return col
+}
+
+func (p *ANSIMeter) Start(label string, total float64) {
+	p.label = []rune(label)
+	p.total = total
+	p.t0 = time.Now().UTC()
+	fmt.Print(cursorInvisible)
+}
+
+func norm(col int, msg []rune) []rune {
+	out := make([]rune, col)
+	copy(out, msg)
+	d := col - len(msg)
+	if d < 0 {
+		out[col-1] = '…'
+	} else {
+		for i := len(msg); i < col; i++ {
+			out[i] = ' '
+		}
+	}
+	return out
+}
+
+func (p *ANSIMeter) SetTotal(total float64) {
+	p.total = total
+}
+
+func (p *ANSIMeter) percent() string {
+	if p.total == 0. {
+		return "---%"
+	}
+	q := p.written * 100 / p.total
+	if q > 999.4 || q < 0. {
+		return "???%"
+	}
+	return fmt.Sprintf("%3.0f%%", q)
+}
+
+func (p *ANSIMeter) Set(current float64) {
```

Comment:
I wonder if we should have some more tests for ansimeter? It seems like it not having many tests currently.

Reply:
:scream: no tests you say?!?

darn, you're probably right.